### DISCUSSION TITLE
Is there a way pseudoTerminals can be made to survive window reloads?

### DISCUSSION THREAD

**Discussion by PEZ** [microsoft/vscode-discussions#199] [Extension Development QnA] [answered]
Regular terminal sessions, included those started by tasks, survive a VS Code window reload. Which is fantastic. In Calva we use a `vscode.Pseudoterminal` for starting an important process. Is there a way we can make this terminal survive the reload?

I haven't investigated this very much. I want to know if it is even supposed to work before I start trying to make it happen. 😀

**Comment by PEZ** [comment]
@Tyriar, in https://github.com/microsoft/vscode/issues/123805#issuecomment-846007351 you say:

>  I think the plan was to make an extension API to opt out of reconnect for _extension owned_ terminals (it's always disabled using `Pseudoterminal` terminals)

Does this mean that I can't have my terminal reconnected? I create it like so:

``` typescript
    jackInTerminal = (<any>vscode.window).createTerminal({
      name: `Calva Jack-in: ${connectSequence.name}`,
      pty: jackInPTY
    });
```

I just quickly tried to add `isTransient: false` there, with no luck, but it could also be that my current cleanup is causing it to close.

**Reply by PEZ** [reply]
Indeed, if I don't provide my pseudoTerminal the terminal survives a reload.

So, I'm out of luck?

**Comment by isidorn** [comment]
Maybe @tyriar can help

**Comment by Tyriar** [comment, marked as answer]
@PEZ since Pseudoterminal-based terminals are connected to the extension host which gets taken down, this is not possible. You could do it on the extension's end manually though by tracking the state of the terminal so that after a relaunch you can re-create the Pseudoterminal and rehydrate it with the data. I don't think we would be able to support this on our end, unless I'm misunderstanding something?

**Reply by PEZ** [reply]
Thanks!

Thing is, that my pseudo terminal actually keeps track of a shell process. It's this process I want to survive. The contents of the pseudo terminal is not as important, but would be good Ux to keep that as well, of course.

The reason I am using a pseudo terminal, and not a regular terminal, is that I need access to the process output. Afaiu, regular terminals won't give me that.

Can you see anyway I can have my process controlled by a reload-surviving terminal where my extension can react on the process output?

**Reply by Tyriar** [reply]
If you manage the shell process it's all on you to re-implement process survival (detach from extension host, allow reconnection somehow), that's the difference between a regular terminal and a `Pseudoterminal`-based one.

> The reason I am using a pseudo terminal, and not a regular terminal, is that I need access to the process output

We have considered such an API in the past but are concerned about performance and security concerns. There is https://github.com/microsoft/vscode/issues/145234 which will revisit this, but in the meantime I don't think you can use a regular terminal.

**Reply by PEZ** [reply]
Thanks again. I think I'll wait a bit more and see if I can avoid re-implementing that. Seems you are closer to add that API to the terminals than you were three years ago. 😄 https://github.com/microsoft/vscode/issues/59337#issuecomment-518965247
